src/gadgets/sumcheck.rs

```diff
@@ -0,0 +1,336 @@
+#![allow(unused)]
+use crate::traits::Group;
+use bellperson::{
+  gadgets::{
+    boolean::{AllocatedBit, Boolean},
+    num::{AllocatedNum, Num},
+    Assignment,
+  },
+  ConstraintSystem, LinearCombination, SynthesisError,
+};
+use ff::{Field, PrimeField, PrimeFieldBits};
+
+pub fn verify<Scalar, CS>(
+  mut cs: CS,
+  claim: AllocatedNum<Scalar>,
+  num_rounds: usize,
+  degree_bound: usize,
+  polys: Vec<Vec<AllocatedNum<Scalar>>>,
+  challenges: Vec<AllocatedNum<Scalar>>, // this is while Transcript circuit is not ready
+) -> Result<(AllocatedNum<Scalar>, Vec<AllocatedNum<Scalar>>), SynthesisError>
+where
+  Scalar: PrimeField + PrimeFieldBits,
+  CS: ConstraintSystem<Scalar>,
+{
+  let mut e = claim;
+  for (i, poly) in polys.iter().enumerate() {
+    // verify poly degree // TODO move check to constraints
+    if poly.len() - 1 != degree_bound {
+      panic!(
+        "poly degree {:?} != degree_bound {:?}",
+        poly.len() - 1,
+        degree_bound
+      );
+    }
+
+    // eval at zero
+    let eval_0 = poly[0].clone();
+    // eval at one
+    let eval_1 = eval_at_one(cs.namespace(|| format!("sc_eval_at_one {}", i)), poly)?;
+
+    cs.enforce(
+      || format!("(s(0)+s(1)) * 1 = s(r) {}", i),
+      |lc| lc + eval_0.get_variable() + eval_1.get_variable(),
+      |lc| lc + CS::one(),
+      |lc| lc + e.get_variable(),
+    );
+
+    let r_i = challenges[i].clone(); // this is while Transcript circuit is not ready
+
+    e = uni_evaluate(
+      cs.namespace(|| format!("sc_unieval {}", i)),
+      poly.clone(),
+      r_i,
+    )?;
+  }
+
+  // note: challenges will be replaced by r vector once the Transcript circuit is ready
+  Ok((e, challenges))
+}
+
+pub fn eval_at_one<Scalar, CS>(
+  mut cs: CS,
+  poly: &Vec<AllocatedNum<Scalar>>,
+) -> Result<AllocatedNum<Scalar>, SynthesisError>
+where
+  Scalar: PrimeField + PrimeFieldBits,
+  CS: ConstraintSystem<Scalar>,
+{
+  let mut eval: Num<Scalar> = Num::<Scalar>::from(poly[0].clone());
+  for (i, coeff) in poly.iter().enumerate().skip(1) {
+    // logic: eval_aux = eval + coef
+    let eval_aux = eval.clone().add(&Num::<Scalar>::from(coeff.clone()));
+    // r1cs constr: (eval + coeff) * 1 = eval_aux
+    let allocated_eval: AllocatedNum<Scalar> =
+      AllocatedNum::<Scalar>::alloc(&mut cs, || Ok(eval.get_value().unwrap()))?;
+    let allocated_eval_aux: AllocatedNum<Scalar> =
+      AllocatedNum::<Scalar>::alloc(&mut cs, || Ok(eval_aux.get_value().unwrap()))?;
+    cs.enforce(
+      || format!("eval + coeff = eval_aux{}", i),
+      |lc| lc + allocated_eval.get_variable() + coeff.get_variable(),
+      |lc| lc + CS::one(),
+      |lc| lc + allocated_eval_aux.get_variable(),
+    );
+    eval = eval_aux;
+  }
+  let allocated_eval: AllocatedNum<Scalar> =
```

Comment:
Can't you copy constrain the value at least? So that there's no way to cheat values here?

Reply:
What do you mean by 'copy constrain'? (in the context of R1CS)

Reply:
What I mean is that we need a way to ensure that the inner loop allocated value is the same as the one created later. 
Otherwise, there's an equality not being checked.